docs/proposals/19-monitoring.md

```diff
@@ -0,0 +1,373 @@
+                summary: Cloud controller manager is down.
+          observedComponents: |
+            observedPods:
+            - podPrefix: cloud-controller-manager
+            isExposedToUser: true
+          scrape_config: |
+            - job_name: cloud-controller-manager
+              scheme: https
+              tls_config:
+                insecure_skip_verify: true
+              authorization:
+                type: Bearer
+                credentials_file: /var/run/secrets/gardener.cloud/shoot/token/token
+              honor_labels: false
+              kubernetes_sd_configs:
+              - role: endpoints
+                namespaces:
+                  names: [shoot--project--name]
+              relabel_configs:
+              - source_labels:
+                - __meta_kubernetes_service_name
+                - __meta_kubernetes_endpoint_port_name
+                action: keep
+                regex: cloud-controller-manager;metrics
+              # common metrics
+              - action: labelmap
+                  regex: __meta_kubernetes_service_label_(.+)
+              - source_labels: [ __meta_kubernetes_pod_name ]
+                  target_label: pod
+              metric_relabel_configs:
+              - source_labels: [ __name__ ]
+                regex: ^(rest_client_requests_total|process_max_fds|process_open_fds)$
+                action: keep
+        ```
+
+    - This config map will be split up into 2 separate resources. One for the
+      `alerting_rules` and another for the `scrape_config`. The `alerting_rules`
+      should be converted into a `prometheusrules` object. Since the
+      `scrape_config` only has one `job_name` we will only need one
+      `servicemonitor` or `podmonitor` for this.
+
+    - Adding meta monitoring for the extensions: If the extensions need to be
+      scraped for monitoring, the extensions must bring their own [Custom
+      Resources][prom-crds].
+        - Currently the contract between extensions and gardener is that
+          anything that needs to be scraped must have the labels:
+          `prometheus.io/scrape=true` and `prometheus.io/port=<port>`. This is
+          defined [here][prom-config]. This is something that we can still
+          support with a `PodMonitor` that will scrape any pod in a specified
+          namespace with these labels.
+
+5. Add a [sidecar][grafana-sidecar] to grafana that will pickup dashboards and
+   provision them. Each dashboard gets its own configmap.
+
+   - Some of the dashboards were initially inspired by [kubernetes-mixins].
+     Gardener should reuse dashboards and alerting configs whenever possible. It
+     is not clear yet how to integrate this into Gardener.
+
+6. Migrating to the new monitoring stack:
+    1. Deploy the [prometheus-operator] and its custom resources.
+    1. Add a feature gate to optionally deploy the custom resources defined in the Proposal.
```

Comment:
Do we really need this feature gate? We should be conscious about this since this comes with quite some effort w.r.t. promotion etc.

Reply:
I was concerned about a graceful migration and figured a feature gate would be the easiest solution. What would your suggestion be? Skip the feature gates and have both stacks be deployed for 1-2 releases, then delete the old stack?

Reply:
Just skip the one for optionally deploying the custom resources. The other one is needed I think since the extensions must be adapted as well.